Allow removing workspaces from the "recent projects" modal

tunght opened the pull request:
Release Notes:

- Delete open projects from the list ([#7426](https://github.com/zed-industries/zed/issues/7426)).

<img width="492" alt="Screenshot 2024-02-19 at 10 59 01 AM" src="https://github.com/zed-industries/zed/assets/1823955/922117f6-81c1-409d-938a-131bcec0f24c">

<img width="675" alt="Screenshot 2024-02-19 at 10 59 27 AM" src="https://github.com/zed-industries/zed/assets/1823955/fefac68b-9a99-43bb-ac0c-724e7c622455">

tunght committed:
Added ability to remove a project from the recent list

tunght committed:
Merge branch 'main' into issue-7426

tunght committed:
Clean up

SomeoneToIgnore commented on crates/recent_projects/src/recent_projects.rs:
```diff
@@ -117,7 +115,7 @@ impl Render for RecentProjects {
 pub struct RecentProjectsDelegate {
     workspace: WeakView<Workspace>,
-    workspace_locations: Vec<WorkspaceLocation>,
+    workspace_locations: Vec<(WorkspaceId, WorkspaceLocation)>,
```
NIT: those seem to be not locations anymore, but just the `workspaces`?

tunght replied:
Makes sense, just a little bit confused with the workspace property above which is a Weakview

tunght committed:
change delete icon to "x"

bennetbo commented:
Are you still working on this @tunght? 
I feel like the delete project button is too easy to accidentally press when selecting a project. What do you think about hiding it inside a context menu, similar to how it's done in the chat panel?
![image](https://github.com/zed-industries/zed/assets/53836821/428a00e1-608e-43e7-a99f-078a176e6256)

tunght commented:
@bennetbo  Yes, Im still working on this. Im still having some issues, and trying to find a way to address them.
I can try to add it in a context menu, the delete button is small and on the far right side thou.
I added a "Delete all" button as well, but I think we need a confirm dialog before actually clearing out all of the projects

Couple of issues I'm having if you can help show me to the right direction that would be great! ( I'm new to the project ):
- Still not sure how to do keybindings yet. I added new key context RecentProjects and new set of actions, however it doesn't work.

```rust
// recent_projects.rs
actions!(recent_projects, [Delete, DeleteAll]);
```
```rust
// recent_projects.rs
impl PickerDelegate for RecentProjectsDelegate {
  fn render_footer(&self, cx: &mut ViewContext<Picker<Self>>) -> Option<AnyElement> {
        if self.workspace_locations.len() == 0 {
            None
        } else {
            Some(
                div()
                    .key_context("RecentProjects")
                    .size_full()
                    .items_center()
                    .on_action(cx.listener(|_this, _: &DeleteAll, _cx| {
                        // this.delegate.delete_all_recent_projects(cx)
                        println!("Deleting all");
                    }))
                    .child(Divider::horizontal())
                    .child(
                        div().w_full().py_1().child(
                            Button::new("clear_all", "Clear All...")
                                .icon_color(Color::Muted)
                                .icon(IconName::Trash)
                                .icon_position(IconPosition::Start)
                                .style(ButtonStyle::Filled)
                                .full_width()
                                .on_click(cx.listener(|this, _, cx| {
                                    this.delegate.delete_all_recent_projects(cx)
                                })),
                        ),
                    )
                    .into_any_element(),
            )
        }
    }
}
```
```json
// assets/keymaps/default.json
{
    "context": "RecentProjects",
    "bindings": {
      "cmd-d": "recent_projects::DeleteAll"
    }
  }
```

SomeoneToIgnore commented:
My thoughts (feel free to ignore and follow your path, if needed): I do not think people will removing recent projects that much, only certain accidental ones, so having all things mentioned above does not make much sense to me in that context.

In particular:
* I would totally ignore the "delete all" case as it is somewhat odd — we do not delete all entries due to certain workspace(s) being open + it adds more complications with modals & other
* Keybindings are nice to have, but are those really necessary? Seems like a complication also.
* Context menu is not really needed too


Here's what VSCode does and I think it's just enough:
![image](https://github.com/zed-industries/zed/assets/2690773/3958f99e-f8d7-40cf-9d6a-e2514b19d4a7)

tunght commented:
It makes sense to me, I never really do the "Delete all", in fact I really don't want to do it. For delete a project, I do sometimes want to do it for duplication, secret projects, ...

tunght commented:
BTW for the button hover, i think it's does actually work, it's just the background of the icon button and the list item are in the same color ( as technically they are both hovered at the same time )

tunght committed:
Only show delete button when hover

tunght committed:
disable delete recent project for current workspace

tunght committed:
Updated variable name

SomeoneToIgnore reviewed (commented):
Looks really great, thank you for picking the minimalistic approach.

Two notes:

* when the list is large enough and I want to remove something from the bottom part where I have to scroll first — the entire list jumps back to top after clicking the X button.
This is somewhat disturbing, could we instead preserve the scroll? I think scrolling to previous element would be good enough.

* we [did discuss](https://github.com/zed-industries/zed/issues/7426#issuecomment-1948275529) that we do not want to remove any open workspace, but the code only does that for the current workspace and ignores other windows.
Were there any issues accessing `WorkspaceStore` in the code?

tunght commented:
Thanks for your feedback! I can fix the scroll and the X button visibility.
The workspace_store seems hard. It took me sometimes and i haven't been success. 

I have found a bug, I didn't know why yet, but still trying to debug to find the exact root cause, but it has to do with deleting workspace. I think after deleting a workspace in the workspaces table, there are still other records related to the workspace in other tables like panes, pane_groups, items, editors...
Now whenever I open a new project, it will insert a new empty row to the workspaces table:
![Screenshot 2024-02-20 at 9 34 16 AM](https://github.com/zed-industries/zed/assets/1823955/2f25aaec-43c9-411c-965d-7b073c579b76)
if I close zed completely and open zed again, unless i open one of the deleted project again, it's working fine. But the moment i open one of the deleted project, then after that it will only insert null row whenever i open any new project

SomeoneToIgnore commented:
>  The workspace_store seems hard. It took me sometimes and i haven't been success.

Had tried it myself and see what you mean now — it's either a circular reference panic (which is right, since we hold to a specific workspace during that) or no window context for the workspace being registered (odd, but does not help us).

So let's ignore that case for now entirely.

-------

Great find with the wrong entries inserted but one question — is this actually a problem of a current PR or, rather, something that you can reproduce in `main` too?
I've failed to get those persistence errors, but did manage to get `null` entries in the DB on `main` when opened zed, closed zed, moved the project, opened zed.

It is bad, but if you can also reproduce it on `main`, let's do one step at a time: add the X on selection here and merge this; create a separate issue with reproduction steps on the DB issue and fix that (if you are willing to, I can try to do both after you confirm that it is an issue on `main` too)

SomeoneToIgnore commented:
Oh, I've actually just stumbled on another way to get all open workspaces, I think: https://github.com/zed-industries/zed/pull/8086/files#diff-89af0b4072205c53b518aa977d6be48997e1a51fa4dbf06c7ddd1fec99fc510eL415-L426

Worth a try but not a blocker if it would not work too.

tunght commented:
I will try that route and check if I can reproduce the issue on main.
Although, I think it's might be better we separate recent open projects out of workspaces table, and only update the recent projects at the opening time.

tunght committed:
Show x button on selection

tunght commented:
@SomeoneToIgnore it's now showing the x on item selection. For scrolling to the item before the deleted item, it would requires to modify the picker public function which might affect other places that use picker.
Right now there are two way to re-render the picker is to use picker.refresh or picker.update_matches. Both functions reset the selected_match_index and it scrolls to the selected_match_index. If we modify the function, it would affect the functionality of the filter.
If it's not a show blocker for this PR, can we go ahead and merge first and create separate issues for the scrolling and the database bug?

tunght committed:
Preserve scroll after deleting an item

tunght commented:
Never mind, I managed to fix the scrolling issue!

tunght committed:
Merge branch 'main' into issue-7426

tunght commented:
About the null and persistent error, I can't replicate it in main. I merged main to this branch and cannot replicate it either.
I did your steps: open zed, close zed, move proj, open zed, it's doing ok now.
But i managed to find another issue both on main and this branch
open proj1, open proj2. Keep zed open and move proj2, open proj3, go to proj2 windows, open recent projects pop up, it will remove proj3 from the recent list

CurbaiCode commented:
Just thought I would mention, in #5357, @iamnbutler's idea used a new UI element—similar to what @SomeoneToIgnore suggested—for removing collaborators:

> **Split context menu items** <img alt="CleanShot - 2022-06-14 at 15 47 58@2x" width="595" src="https://user-images.githubusercontent.com/1714999/173675829-9b3df0a7-23a4-4685-8629-d811bd377846.png">

> <img alt="CleanShot - 2022-06-14 at 15 47 01@2x" width="1385" src="https://user-images.githubusercontent.com/1714999/173675720-e8d1f685-756f-48f9-a65a-9a9cba70b936.png">

SomeoneToIgnore reviewed (approved):
Beautiful, thank you for fixing the scroll — with that in mind, I think we can ignore the persistence issues for now (those are supposed to get a rewrite anyway, as I've learned discussing this PR offline) and ignore the multi-workspace issue too.

Whatever else bugs, let's move on for now: I failed to repro the case you've mentioned in the recent comment and that seems rather hard to get into, compared with the general feature usefulness.

Could you please rebase on top of `main` to include https://github.com/zed-industries/zed/pull/8154 to fix the CI?

tunght committed:
Merge branch 'main' into issue-7426

tunght commented:
Awesome, thank you! I updated the branch with latest from main
